internal/query/query.go

```diff
@@ -6,97 +6,19 @@ import (
 
 	"github.com/graph-gophers/graphql-go/errors"
 	"github.com/graph-gophers/graphql-go/internal/common"
+	"github.com/graph-gophers/graphql-go/types"
 )
 
-type Document struct {
-	Operations OperationList
-	Fragments  FragmentList
-}
-
-type OperationList []*Operation
-
-func (l OperationList) Get(name string) *Operation {
-	for _, f := range l {
-		if f.Name.Name == name {
-			return f
-		}
-	}
-	return nil
-}
-
-type FragmentList []*FragmentDecl
-
-func (l FragmentList) Get(name string) *FragmentDecl {
-	for _, f := range l {
-		if f.Name.Name == name {
-			return f
-		}
-	}
-	return nil
-}
-
-type Operation struct {
-	Type       OperationType
-	Name       common.Ident
-	Vars       common.InputValueList
-	Selections []Selection
-	Directives common.DirectiveList
-	Loc        errors.Location
-}
-
-type OperationType string
-
 const (
-	Query        OperationType = "QUERY"
-	Mutation                   = "MUTATION"
-	Subscription               = "SUBSCRIPTION"
+	Query        types.OperationType = "QUERY"
+	Mutation                         = "MUTATION"
+	Subscription                     = "SUBSCRIPTION"
 )
 
-type Fragment struct {
-	On         common.TypeName
-	Selections []Selection
-}
-
-type FragmentDecl struct {
-	Fragment
-	Name       common.Ident
-	Directives common.DirectiveList
-	Loc        errors.Location
-}
-
-type Selection interface {
-	isSelection()
-}
-
-type Field struct {
-	Alias           common.Ident
-	Name            common.Ident
-	Arguments       common.ArgumentList
-	Directives      common.DirectiveList
-	Selections      []Selection
-	SelectionSetLoc errors.Location
-}
-
-type InlineFragment struct {
-	Fragment
-	Directives common.DirectiveList
-	Loc        errors.Location
-}
-
-type FragmentSpread struct {
-	Name       common.Ident
-	Directives common.DirectiveList
-	Loc        errors.Location
-}
-
-func (Field) isSelection()          {}
-func (InlineFragment) isSelection() {}
-func (FragmentSpread) isSelection() {}
-
-func Parse(queryString string) (*Document, *errors.QueryError) {
+func Parse(queryString string) (*types.ExecutableDefinition, *errors.QueryError) {
 	l := common.NewLexer(queryString, false)
 
-	var doc *Document
+	var doc *types.ExecutableDefinition
```

Comment:
Why is `Document` replaced with `ExecutableDefinition`?
As the spec says a document might include other things other than just `ExecutableDefinition`,  no?

> A GraphQL Document describes a complete file or request string operated on by a GraphQL service or client. 

Although, it also says that:

> GraphQL services which only seek to provide GraphQL query execution may choose to only include ExecutableDefinition and omit the TypeSystemDefinition and TypeSystemExtension rules from Definition.

The question is do we want to exclude `TypeSystemDefinition` and `TypeSystemExtension`?

Reply:
We felt that the type called Document did not match up with the specification of a Document. As you mention, a document may contain an `ExecutableDefinition` but also a `TypeSystemDefinition` and extensions. It's essentially the root type. The grammar describes this here: http://spec.graphql.org/June2018/#sec-Appendix-Grammar-Summary.Document and the type itself is documented here: https://github.com/graph-gophers/graphql-go/pull/437/files/130895f516efd0fc6c23dca24bcdf2a8cf0c24bf#diff-b349fbf81a907441c5bb68aa1c730e71b6492354cd594b63626d781dc0bd51e7R9

Rather than add more things to Document and refactor everything to work based on that root type, we thought that making this type more narrow (in name only) was a better choice as it makes this change a bit more surgical and less risky. Another advantage of this naming choice is that the name `types.Document` is free to use as the root of the AST to match the grammar and specification. I'm open to making this change, but it would mean changing a lot of code and likely retiring `types.Schema` and `schema.Schema`.

I noticed that the variable name (`doc`) did not get renamed along with the type. That's an oversight and I'll do a pass to check for this type of error later on today.

Reply:
> I noticed that the variable name (doc) did not get renamed along with the type. That's an oversight and I'll do a pass to check for this type of error later on today.

That would be idea, to avoid confusion in future.

Reply:
Just one question re:
> Rather than add more things to Document and refactor everything to work based on that root type
Why do we need to add everything and why is refactoring required? Do you mean adding `ExecutableDefinition` under document?

Reply:
> Why do we need to add everything and why is refactoring required? Do you mean adding ExecutableDefinition under document?

I think we have some crossed wires, so I'm going to try and re-establish my line of thinking.

It's my belief that a type called `types.Document` should match the specification for a Document as closely as is reasonable. The type called `Document` in the query package did not match the specification, as it was missing 2/3 of the fields. http://spec.graphql.org/June2018/

The type now called `types.ExecutableDefinition` matches the relevant part of the specification. It's clear to everyone what it does and which part of the specification it implements.

An orthogonal option available to us in this PR (or a followup) is to introduce this type:

```go
// http://spec.graphql.org/June2018/#sec-Language.Document
type Document struct {
  ExecutableDefinitions []ExecutableDefinition
  TypeSystemDefinitions []TypeSystemDefinition
  TypeSystemExtension []TypeSystemExtension
}
```

and

```
// http://spec.graphql.org/June2018/#TypeSystemDefinition
type TypeSystemDefinition struct {
  SchemaDefinition SchemaDefinition
  TypeDefinition       []TypeDefinition
  DirectiveDefinition []DirectiveDefinition
}
```

and so on.

The main work of this approach would be to replace references to `types.Schema` with `types.Document` or `types.TypeSystemDefinition`. Given how much code depends on `types.Schema` and its fields, I don't find this option very appealing but it is the most correct way to proceed. I'm interested to hear your thoughts on this.

Reply:
@pavelnikolov any thoughts here? I feel pretty strongly that we shouldn't have a type called `Document` in the public API that doesn't match the specification.

Reply:
I agree with you. Your reason to use `types.ExecutableDefinition` sounds pretty compelling. Let's proceed with executable definition and come back to Document in a followup PR.